core/src/worker/workflow/wft_extraction.rs

```diff
@@ -0,0 +1,132 @@
+use crate::{
+    abstractions::OwnedMeteredSemPermit,
+    protosext::ValidPollWFTQResponse,
+    worker::{
+        client::WorkerClient,
+        workflow::{
+            history_update::HistoryPaginator, CacheMissFetchReq, HistoryUpdate, NextPageReq,
+            PermittedWFT,
+        },
+    },
+};
+use futures::Stream;
+use futures_util::{stream, stream::PollNext, FutureExt, StreamExt};
+use std::{future, sync::Arc};
+use tracing::Span;
+
+/// Transforms incoming validated WFTs and history fetching requests into [PermittedWFT]s ready
+/// for application to workflow state
+pub(super) struct WFTExtractor {}
+
+pub(super) enum WFTExtractorOutput {
+    NewWFT(PermittedWFT),
+    FetchResult(PermittedWFT, Arc<HistfetchRC>),
+    NextPage {
+        paginator: HistoryPaginator,
+        update: HistoryUpdate,
+        span: Span,
+        rc: Arc<HistfetchRC>,
+    },
+    FailedFetch {
+        run_id: String,
+        err: tonic::Status,
+    },
+    PollerDead,
+}
+
+type WFTStreamIn = (
+    Result<ValidPollWFTQResponse, tonic::Status>,
+    OwnedMeteredSemPermit,
+);
+#[derive(derive_more::From, Debug)]
+pub(super) enum HistoryFetchReq {
+    Full(CacheMissFetchReq, Arc<HistfetchRC>),
+    NextPage(NextPageReq, Arc<HistfetchRC>),
+}
+/// Used inside of `Arc`s to ensure we don't shutdown while there are outstanding fetches.
+#[derive(Debug)]
+pub(super) struct HistfetchRC {}
+
+impl WFTExtractor {
+    pub(super) fn build(
+        client: Arc<dyn WorkerClient>,
+        max_fetch_concurrency: usize,
+        wft_stream: impl Stream<Item = WFTStreamIn> + Send + 'static,
+        fetch_stream: impl Stream<Item = HistoryFetchReq> + Send + 'static,
+    ) -> impl Stream<Item = Result<WFTExtractorOutput, tonic::Status>> + Send + 'static {
+        let fetch_client = client.clone();
+        let wft_stream = wft_stream
+            .map(move |(wft, permit)| {
+                let client = client.clone();
+                async move {
+                    match wft {
+                        Ok(wft) => {
+                            let prev_id = wft.previous_started_event_id;
+                            let run_id = wft.workflow_execution.run_id.clone();
```

Comment:
Not without making the code substantially more annoying than this would be worth. Not cloning IDs everywhere should be generally fixed at some point anyways, to the extent it matters much.